replicator/main.go

```diff
@@ -0,0 +1,847 @@
+
+		totalCount: count,
+	}
+
+	dr.computeRowDeleteQuery()
+	dr.computePartitionDeleteQuery()
+
+	return dr, nil
+}
+
+func (r *DeltaReplicator) computeRowDeleteQuery() {
+	keyColumns := append(append([]string{}, r.pkColumns...), r.ckColumns...)
+
+	r.rowDeleteQueryStr = fmt.Sprintf(
+		"DELETE FROM %s WHERE %s",
+		r.tableName,
+		r.makeBindMarkerAssignments(keyColumns, " AND "),
+	)
+}
+
+func (r *DeltaReplicator) computePartitionDeleteQuery() {
+	r.partitionDeleteQueryStr = fmt.Sprintf(
+		"DELETE FROM %s WHERE %s",
+		r.tableName,
+		r.makeBindMarkerAssignments(r.pkColumns, " AND "),
+	)
+}
+
+func (r *DeltaReplicator) Consume(c scylla_cdc.Change) error {
+	timestamp := c.GetCassandraTimestamp()
+	pos := 0
+
+	if showTimestamps {
+		log.Printf("Processing timestamp: %v\n", c.Time)
+	}
+
+	for pos < len(c.Delta) {
+		change := c.Delta[pos]
+		var err error
+		switch change.GetOperation() {
+		case scylla_cdc.Update:
+			err = r.processUpdate(timestamp, change)
+			pos++
+
+		case scylla_cdc.Insert:
+			err = r.processInsert(timestamp, change)
+			pos++
+
+		case scylla_cdc.RowDelete:
+			err = r.processRowDelete(timestamp, change)
+			pos++
+
+		case scylla_cdc.PartitionDelete:
+			err = r.processPartitionDelete(timestamp, change)
+			pos++
+
+		case scylla_cdc.RangeDeleteStartInclusive, scylla_cdc.RangeDeleteStartExclusive:
+			// TODO: Check that we aren't at the end?
+			start := change
+			end := c.Delta[pos+1]
+			err = r.processRangeDelete(timestamp, start, end)
+			pos += 2
+
+		default:
+			panic("unsupported operation: " + change.GetOperation().String())
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	r.localCount += int64(len(c.Delta))
+
+	return nil
+}
+
+func (r *DeltaReplicator) End() {
+	// TODO: Take a snapshot here
+	atomic.AddInt64(r.totalCount, r.localCount)
+}
+
+func (r *DeltaReplicator) processUpdate(timestamp int64, c *scylla_cdc.ChangeRow) error {
+	return r.processInsertOrUpdate(timestamp, false, c)
+}
+
+func (r *DeltaReplicator) processInsert(timestamp int64, c *scylla_cdc.ChangeRow) error {
+	return r.processInsertOrUpdate(timestamp, true, c)
+}
+
+func (r *DeltaReplicator) processInsertOrUpdate(timestamp int64, isInsert bool, c *scylla_cdc.ChangeRow) error {
+	runQuery := func(q string, vals []interface{}) error {
+		if debugQueries {
+			fmt.Println(q)
+			fmt.Println(vals...)
+		}
+
+		return tryWithExponentialBackoff(func() error {
+			return r.session.
+				Query(q, vals...).
+				Consistency(r.consistency).
+				Idempotent(true).
+				WithTimestamp(timestamp).
+				Exec()
+		})
+	}
+
+	keyColumns := append(r.pkColumns, r.ckColumns...)
+
+	if isInsert {
+		// Insert row to make a row marker
+		// The rest of the columns will be set by using UPDATE queries
+		var bindMarkers []string
+		for _, columnName := range keyColumns {
+			bindMarkers = append(bindMarkers, makeBindMarkerForType(r.columnTypes[columnName]))
+		}
+
+		insertStr := fmt.Sprintf(
+			"INSERT INTO %s (%s) VALUES (%s) USING TTL ?",
```

Comment:
Gocql has a single query cache per session, and this query will be prepared. I'll move construction of this string to `NewDeltaReplicator`, though, so that we won't have to re-create this string over and over again.

Reply:
ahh so it automatically prepares? Nice

Reply:
Yup, queries are automatically prepared.

Now, the string is constructed once in `precomputeQueries()` and cached inside the `DeltaReplicator` struct.